### PR TITLE
[ws-daemon] Refactor enforcing of CPU limits

### PR DESCRIPTION
**Pull request by aledbf**
## Description

- Remove `jiffies` concept
- Rely on `cpu.cfs_period_us`

## How to test

- Open a workspace
- Download cpuburn `wget https://cdn.pmylund.com/files/tools/cpuburn/linux/cpuburn-1.0-amd64.tar.gz && tar zxpvf cpuburn-1.0-amd64.tar.gz`
- Start CPU burn `./cpuburn/cpuburn`
- After 30s check ws-daemon logs
- The log should contain an output similar to the next one

```
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":-1,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":500,"message":"set new CPU limit","quota":500000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:11:12Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":500000,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":400,"message":"set new CPU limit","quota":400000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:11:42Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":400000,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":200,"message":"set new CPU limit","quota":200000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:12:12Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","instanceId":"3214228e-a0eb-4b6a-91eb-c34bb7747028","level":"debug","message":"the quota has not changed","quota":200000,"serviceContext":{"service":"ws-daemon","version":"commit-b7b4c81e4c98efecca85740b0e07a2ebf79261f8"},"severity":"DEBUG","time":"2022-01-25T00:21:39Z","userId":"","workspaceId":"gitpodio-templatetypescr-loh4kwdzby1"}

```

## Release Notes
```release-note
[ws-daemon] Refactor enforcing of CPU limits
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **aledbf** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/ws-daemon/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/ws-daemon/pkg/resources/controller_test.go [hunk 1]
```diff
@@ -54,43 +54,43 @@ func TestControlCPU(t *testing.T) {
 			Name:      "bucket regular use",
 			Opts:      []ControllerOpt{WithCPULimiter(bktLimiter())},
 			Consumer:  splitpointConsumer(fixedConsumer(500), randomConsumer(20, 100), 5*time.Minute),
-			Validator: aucValidator(49985),
+			Validator: aucValidator(53385),
```
Comment (csweichel) [inline components/ws-daemon/pkg/resources/controller_test.go:57]: Assuming the period stays the same in our tests (100ms), and the units remain the same (mCPU), why do those numbers change?

Comment (aledbf) [reply components/ws-daemon/pkg/resources/controller_test.go]: fixed

File: components/ws-daemon/pkg/resources/dispatch.go [hunk 2]
```diff
@@ -84,14 +84,11 @@ func (d *DispatchListener) WorkspaceAdded(ctx context.Context, ws *dispatch.Work
 
 	var cpuLimiter ResourceLimiter
 	if fixedLimit, ok := ws.Pod.Annotations[wsk8s.CPULimitAnnotation]; ok && fixedLimit != "" {
-		var scaledLimit int64
-		limit, err := resource.ParseQuantity(fixedLimit)
+		scaledLimit, err := resource.ParseQuantity(fixedLimit)
 		if err != nil {
 			log.WithError(err).WithField("limitReq", fixedLimit).Warn("workspace requested a fixed CPU limit, but we cannot parse the value")
```
[the remark this answers is not in the document]
Comment (aledbf) [reply components/ws-daemon/pkg/resources/dispatch.go:89]: fixed

**Comment by sagor999** [comment]
@csweichel this seems to be working correctly. 
I just ran a test. Updated ws-daemon configmap with  budget map that allocated 600s of work per bucket.
Then ran workspace and ran cpu burn in there.
https://github.com/gitpod-io/gitpod/blob/5459604e982a58bed01ae31024ca66db6e537db2/components/ws-daemon/pkg/resources/controller.go#L209
for each 10 seconds, it was logging 50seconds of work done (5 cpu each doing 10s of work)
after 600s of work done, new cpu limit was applied, limiting to 4 cpu. now for each 10 seconds it was reporting 40 seconds of work. Then next and last limit, for each 10s was reporting 20s of work.
Could also see similar limiting when running `htop` in that workspace.

Could you try again? 
if it works for you, I will probably just update logging to be less verbose (switch it to debug). Unless there is something else missing in this PR?

Also I assume this is by design(?), but limits persists per workspace. So if I stop workspace and start a new one, I start fresh.

**Comment by sagor999** [comment]
Closing this PR as it is superseded by this one: #8036

**Pull request closed without merging**